### PR TITLE
SWATCH-951, SWATCH-952: swatch-contracts GET/POST/PUT api corrections

### PR DESCRIPTION
**Pull request by lindseyburnett**
https://issues.redhat.com/browse/SWATCH-951
https://issues.redhat.com/browse/SWATCH-952

test http file & curl commands

```
# curl -X 'POST'
#  'http://localhost:8000/api/swatch-contracts/internal/contracts'
#  -H 'accept: application/json'
#  -H 'Content-Type: application/json'
#  -d '{
#  "subscription_number": "sn789",
#  "sku": "MW999",
#  "start_date": "2023-03-07T00:00:00Z",
#  "org_id": "RACEMS",
#  "billing_provider": "aws",
#  "billing_account_id": "bp1",
#  "product_id": "BASILISK",
#  "metrics": [
#    {
#      "metric_id": "instance-hours",
#      "value": 100
#    }
#  ]
#}'
POST http://localhost:8000/api/swatch-contracts/internal/contracts
accept: application/json
Content-Type: application/json

{
  "subscription_number": "AAsn78934",
  "sku": "MW999",
  "start_date": "2023-03-07T00:00:00Z",
  "org_id": "RACEMS",
  "billing_provider": "aws",
  "billing_account_id": "bp1",
  "product_id": "BASILISK",
  "metrics": [
    {
      "metric_id": "instance-hours",
      "value": 100
    }
  ]
}

###

# curl -X 'GET'
#  'http://localhost:8000/api/swatch-contracts/internal/contracts'
#  -H 'accept: application/json'
GET http://localhost:8000/api/swatch-contracts/internal/contracts?subscriptionNumber=sn78934
accept: application/json

###

# curl -X 'PUT'
#  'http://localhost:8000/api/swatch-contracts/internal/contracts/b176a3a2-ab44-4645-9399-90f4c0c22613'
#  -H 'accept: application/json'
#  -H 'Content-Type: application/json'
#  -d '{
#  "billing_account_id": "bp1",
#  "billing_provider": "aws",
#  "metrics": [
#    {
#      "metric_id": "instance-hours",
#      "value": 100
#    }
#  ],
#  "org_id": "RACEMS",
#  "product_id": "BASILISK",
#  "sku": "MW999",
#  "start_date": "2023-03-07T00:00:00Z",
#  "subscription_number": "sn78934",
#    "uuid": "b176a3a2-ab44-4645-9399-90f4c0c22613"
#}'
PUT http://localhost:8000/api/swatch-contracts/internal/contracts/b176a3a2-ab44-4645-9399-90f4c0c22613
accept: application/json
Content-Type: application/json

{
  "billing_account_id": "bp1",
  "billing_provider": "aws",
  "metrics": [
    {
      "metric_id": "instance-hours",
      "value": 1000
    }
  ],
  "org_id": "RACEMS",
  "product_id": "BASILISK",
  "sku": "MW999",
  "start_date": "2023-03-07T00:00:00Z",
  "subscription_number": "sn789341111",
  "uuid": "b176a3a2-ab44-4645-9399-90f4c0c22613"
}

###

# curl -X 'GET'
#  'http://localhost:8000/api/swatch-contracts/internal/contracts?org_id=RACEMS'
#  -H 'accept: application/json'
GET http://localhost:8000/api/swatch-contracts/internal/contracts?org_id=RACEMS
accept: application/json

###

```

### CODE REVIEW

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/ContractRepository.java [hunk 1]
```diff
@@ -20,97 +20,16 @@
...
-              + " m on c.uuid = m.contractUuid where ";
-      query = metricTableJoin + query;
-    }
-
-    log.info("Dynamically generated query: {}", query);
-
-    return find(query, nonNullParams).list();
-  }
-
-  public Optional<ContractEntity> getContract(
```
Comment (kartikshahc) [inline swatch-contracts/src/main/java/com/redhat/swatch/contract/repository/ContractRepository.java:84]: Before you remove this method could we make sure we have equivalent specification method?

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java [hunk 2]
```diff
@@ -207,10 +244,10 @@ private void persistContract(ContractEntity entity, OffsetDateTime now) {
   }
 
   private Optional<ContractEntity> currentlyActiveContract(ContractEntity contract) {
-    Map<String, Object> stringObjectMap =
-        Map.of("subscriptionNumber", contract.getSubscriptionNumber());
-
-    return contractRepository.getContract(stringObjectMap, true);
```
Comment (kartikshahc) [inline swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java:213]: This isn't duplicate code. :)
I guess you need multiple contracts but, for upstream I need single contract by subscriptionnumber and end date null. Before we delete this lets make sure that is accounted.

File: swatch-contracts/src/test/java/com/redhat/swatch/contract/service/ContractServiceTest.java [hunk 3]
```diff
@@ -263,10 +261,7 @@ void createPartnerContract_NotDuplicateContractThenPersist() {
     productTags.data(List.of("MH123"));
     when(syncResource.getSkuProductTags(any())).thenReturn(productTags);
 
-    Map<String, Object> stringObjectMap =
-        Map.of("subscriptionNumber", contract.getRedHatSubscriptionNumber());
-    when(contractRepository.getContract(stringObjectMap, true))
-        .thenReturn(Optional.of(existingContract));
+    when(contractRepository.getContracts(any())).thenReturn(List.of(existingContract));
```
Comment (kartikshahc) [inline swatch-contracts/src/test/java/com/redhat/swatch/contract/service/ContractServiceTest.java:264]: This is for partner contract where we need to get only single contract as explained above. Would like to see its equivalent if we are doing based on Specification.

File: swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java (hunk 2 above)
Comment (lindseyburnett) [reply swatch-contracts/src/main/java/com/redhat/swatch/contract/service/ContractService.java:213]: I think we have the equivalent here:

https://github.com/RedHatInsights/rhsm-subscriptions/pull/1930/files/f6f6c460c7532f1284bb8f43ca3eec30ba4641bc#diff-d0a3f94bef59f416c81d8ea1f761cdfcd6e5ae3e5888e580e5b398772ce8b5f1R247

**Review by kahowell — CHANGES_REQUESTED** [review]
Overall, I'm happy but I noticed a small mistake where a message was removed. I'll just add a commit to fix this, since fix is obvious & trivial.

**Review by kahowell — APPROVED** [review]

**Review by kartikshahc — APPROVED** [review]

**Pull request merged**